File: clients/java/src/main/java/com/hook0/client/DecodeException.java
package com.hook0.client;

/**
 * What the API answered is not what it declares it answers.
 *
 * <p>Reading is deliberately strict. A member the document declares as a string and the API answered as a number stops
 * the read with the name of that member, rather than yielding a value whose documentation lies about what it holds.
 * Every failure of that kind is one of these, so a caller has one thing to catch whatever the shape of the answer was.
 */
public class DecodeException extends Hook0Exception {

  private static final long serialVersionUID = 1L;

  /**
   * Builds one out of what could not be read.
   *
   * @param detail what to say about the failure
   */
  public DecodeException(String detail) {
    super(detail);
  }

  /**
   * Builds one out of what could not be read and what stopped the read.
   *
   * @param detail what to say about the failure
   * @param cause what this failure was raised out of
   */
  public DecodeException(String detail, Throwable cause) {
    super(detail, cause);
  }
}

File: clients/kotlin/src/main/kotlin/com/hook0/kotlin/DecodeException.kt
package com.hook0.kotlin

/**
 * An answer that is not what the API declares it answers.
 *
 * A member the document requires and the API did not send, a string where a number was declared, a
 * value outside a closed list: each of them stops the read here rather than travelling on as a
 * value nobody checked.
 */
class DecodeException : Hook0Exception {

  /**
   * Builds one out of what could not be read.
   *
   * @param detail what to say about the failure
   */
  constructor(detail: String) : super(detail)

  /**
   * Builds one out of what could not be read and what reported it.
   *
   * @param detail what to say about the failure
   * @param cause what this failure was raised out of
   */
  constructor(detail: String, cause: Throwable) : super(detail, cause)
}
